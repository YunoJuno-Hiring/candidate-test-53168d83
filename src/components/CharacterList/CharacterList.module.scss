@import 'styles/_mixins';

.CharacterList-container {
  width: 98vw;

  @include lg {
    width: 60vw;
    min-height: 10vh;
  }
}
