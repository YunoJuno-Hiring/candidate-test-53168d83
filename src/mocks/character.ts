import { ICharacter } from '@custom-types/character';

export function generateMockCharacter(overrides?: Partial<ICharacter>): ICharacter {
  const mockCharacter = {
    name: 'Gandalf the Grey',
    category: 'wizard' as const,
    description:
      'A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.',
    significanceIndex: 1,
    avatar: 'gandalf_the_grey.jpg'
  };

  return { ...mockCharacter, ...overrides };
}
