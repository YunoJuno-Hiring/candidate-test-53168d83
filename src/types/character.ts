import * as strings from '@constants/strings';

const { characterCategoryStrings } = strings;

export type TCharacterCategories =
  (typeof characterCategoryStrings)[keyof typeof characterCategoryStrings];

export interface ICharacter {
  category: TCharacterCategories;
  name: string;
  description: string;
  significanceIndex: number;
  avatar: string;
}
