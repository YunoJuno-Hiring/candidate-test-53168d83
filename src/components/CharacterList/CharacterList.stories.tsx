import React from 'react';

import type { Meta, StoryObj } from '@storybook/react';
import CharacterList from './CharacterList';
import FilterContextProvider from '@contexts/FilterContext';

const meta: Meta<typeof CharacterList> = {
  component: CharacterList,
  decorators: [
    (Story) => (
      <FilterContextProvider>
        <Story />
      </FilterContextProvider>
    )
  ]
};

export default meta;

type Story = StoryObj<typeof CharacterList>;

export const Default: Story = {
  render: () => <CharacterList />
};
