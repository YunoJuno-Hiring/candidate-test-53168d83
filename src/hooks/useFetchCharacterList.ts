import { useFilterContext } from '@contexts/FilterContext';
import { ICharacter } from '@custom-types/character';
import { UseFetchCharactersResponse } from '@custom-types/hooks';
import { handleFilters } from '@utils/filterHelpers';
import { useEffect, useState } from 'react';

export const useFetchCharacterList = (url: string) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [characterList, setCharacterList] = useState<ICharacter[]>([]);
  const { selectedValues } = useFilterContext();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        await fetch(url)
          .then((response) => response.json())
          .then((data) => {
            setCharacterList(handleFilters(data, selectedValues));
            setLoading(false);
          });
      } catch (error) {
        console.log('Error fetching data: ', error);
        setLoading(false);
      }
    };

    fetchData();
  }, [url, selectedValues]);

  return <UseFetchCharactersResponse>{ loading, characterList };
};
