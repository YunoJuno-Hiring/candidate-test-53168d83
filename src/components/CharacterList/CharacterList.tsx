import { CharacterCard } from '@components/CharacterCard';
import { useFetchCharacters } from '@hooks/useFetchCharacters';
import React from 'react';
import { LoadingSpinner } from '@components/LoadingSpinner';
import { Character } from '@custom-types/character';
import { UseFetchCharactersProps } from '@custom-types/hooks';
import styles from './CharacterList.module.scss';

export default function CharacterList() {
  const { loading, characterList }: UseFetchCharactersProps = useFetchCharacters(
    `${process.env.REACT_APP_CHARACTER_LIST_URL}`
  );

  if (loading) return <LoadingSpinner />;

  return (
    <div className={styles['CharacterList-container']}>
      {!loading &&
        characterList &&
        characterList.map((character: Character, key: number) => (
          <CharacterCard characterData={character} key={key} />
        ))}
    </div>
  );
}
