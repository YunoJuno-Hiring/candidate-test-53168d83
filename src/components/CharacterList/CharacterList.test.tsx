import { generateMockCharacter } from '@mocks/character';
import { cleanup, render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

import * as useFetchMock from '@hooks/useFetchCharacterList';
import CharacterList from './CharacterList';

const mockCharacter = generateMockCharacter();

afterEach(() => {
  cleanup();
});

beforeEach(() => {
  process.env.REACT_APP_AVATAR_DIR = '/test-location/';
});

describe('CharacterList component', () => {
  describe('When loading is true', () => {
    it('should return a loading spinner', () => {
      jest.spyOn(useFetchMock, 'useFetchCharacterList').mockReturnValue({
        loading: true,
        characterList: []
      });
      const component = render(<CharacterList />);
      expect(component.getByTestId('loading-spinner')).toBeDefined();
    });
  });

  describe('When loading is false', () => {
    it('should return a list of character cards', () => {
      jest.spyOn(useFetchMock, 'useFetchCharacterList').mockReturnValue({
        loading: false,
        characterList: [mockCharacter]
      });
      const component = render(<CharacterList />);

      expect(component.getByText(mockCharacter.name)).toBeDefined();
    });

    it('Should Match DOM snapshot', () => {
      jest.spyOn(useFetchMock, 'useFetchCharacterList').mockReturnValue({
        loading: false,
        characterList: [mockCharacter, mockCharacter, mockCharacter]
      });

      const component = renderer.create(<CharacterList />).toJSON();
      expect(component).toMatchSnapshot();
    });
  });
});
