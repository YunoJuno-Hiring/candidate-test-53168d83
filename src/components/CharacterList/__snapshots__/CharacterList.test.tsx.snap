// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList component When loading is false Should Match DOM snapshot 1`] = `
<div
  className="CharacterList-container"
>
  <div
    className="CharacterCard"
  >
    <div
      className="CharacterCard-avatar-wrapper"
    >
      <img
        className="CharacterCard-avatar"
        src="undefinedgandalf_the_grey.jpg"
      />
    </div>
    <div
      className="CharacterCard-content-container"
    >
      <p
        className="CharacterCard-content"
      >
        <strong>
          Gandalf the Grey
        </strong>
      </p>
      <p
        className="CharacterCard-content"
      >
        Wizard
      </p>
      <p
        className="CharacterCard-content"
      >
        A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.
      </p>
    </div>
  </div>
  <div
    className="CharacterCard"
  >
    <div
      className="CharacterCard-avatar-wrapper"
    >
      <img
        className="CharacterCard-avatar"
        src="undefinedgandalf_the_grey.jpg"
      />
    </div>
    <div
      className="CharacterCard-content-container"
    >
      <p
        className="CharacterCard-content"
      >
        <strong>
          Gandalf the Grey
        </strong>
      </p>
      <p
        className="CharacterCard-content"
      >
        Wizard
      </p>
      <p
        className="CharacterCard-content"
      >
        A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.
      </p>
    </div>
  </div>
  <div
    className="CharacterCard"
  >
    <div
      className="CharacterCard-avatar-wrapper"
    >
      <img
        className="CharacterCard-avatar"
        src="undefinedgandalf_the_grey.jpg"
      />
    </div>
    <div
      className="CharacterCard-content-container"
    >
      <p
        className="CharacterCard-content"
      >
        <strong>
          Gandalf the Grey
        </strong>
      </p>
      <p
        className="CharacterCard-content"
      >
        Wizard
      </p>
      <p
        className="CharacterCard-content"
      >
        A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.
      </p>
    </div>
  </div>
</div>
`;
